sdk/src/Core/Amazon.Runtime/Internal/Util/Logger.cs

```diff
@@ -80,39 +77,76 @@ private void ConfigureLoggers()
                     il.IsEnabled = (logging & LoggingOptions.Console) == LoggingOptions.Console;
                 if (il is InternalSystemDiagnosticsLogger)
                     il.IsEnabled = (logging & LoggingOptions.SystemDiagnostics) == LoggingOptions.SystemDiagnostics;
+
+                if (!IsEnabled)
+                {
+                    IsEnabled = il.IsEnabled;
+                }
+                if (!IsErrorEnabled)
+                {
+                    IsErrorEnabled = il.IsErrorEnabled;
+                }
+                if (!IsInfoEnabled)
+                {
+                    IsInfoEnabled = il.IsInfoEnabled;
+                }
+                if (!IsDebugEnabled)
+                {
+                    IsDebugEnabled = il.IsDebugEnabled;
+                }
             }
         }
 
         #region Static accessor
 
         public static Logger GetLogger(Type type)
         {
-            if (type == null) throw new ArgumentNullException("type");
+            if (type == null) throw new ArgumentNullException(nameof(type));
 
-            Logger l;
-            lock (cachedLoggers)
-            {
-                if (!cachedLoggers.TryGetValue(type, out l))
-                {
-                    l = new Logger(type);
-                    cachedLoggers[type] = l;
-                }
-            }
-            return l;
+            // Use a lazy initialization to ensure that we only create a logger for a given type once because
+            // the constructor does some heavy lifting including setting up event listeners.
+            var lazyLogger = cachedLoggers.GetOrAdd(type, static t => new Lazy<Logger>(() => new Logger(t)));
+            return lazyLogger.Value;
         }
 
         public static void ClearLoggerCache()
         {
-            lock (cachedLoggers)
+            var newLoggerCache = new ConcurrentDictionary<Type, Lazy<Logger>>();
+            ConcurrentDictionary<Type, Lazy<Logger>> oldLoggerCache;
+
+            do
             {
-                cachedLoggers = new Dictionary<Type, Logger>();
+                oldLoggerCache = cachedLoggers;
+            } while (Interlocked.CompareExchange(ref cachedLoggers, newLoggerCache, oldLoggerCache) != oldLoggerCache);
+
+            // Unregister all the loggers in the old cache
+            foreach (var item in oldLoggerCache ?? Enumerable.Empty<KeyValuePair<Type, Lazy<Logger>>>())
+            {
+                var lazyLogger = item.Value;
+                if (lazyLogger.IsValueCreated)
+                {
+                    lazyLogger.Value.Unregister();
+                }
```

Comment:
I have realized the config change handler is never unregistered and effectively the old logger instances can still be rooted.